codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java

```diff
@@ -772,6 +772,19 @@ void fireChildReadComplete() {
             }
         }
 
+        void fireChildExceptionCaught(Throwable cause) {
+            pipeline().fireExceptionCaught(wrapStreamClosedError(cause));
```

Comment:
It seems wrong to use this wrapping for `fireExceptionCaught`. It doesn't seem "normal" for other Channels to `fireExceptionCaught(ChannelClosedException)`; other channels just fail the futures.

In the case of us receiving a RST_STREAM, I think that should be propagated on the child channel as a userEvent of type Http2ResetFrame. That's already being propagated from the FrameCodec as a read; we should just special-case it in Http2MultiplexCodec.onHttp2Frame. I'm not sure what that means we should do with this exception though, because it seems it _may_ be a duplicate, but that's not guaranteed.

Reply:
@ejona86 I think this brings us back to another "issue" I am currently working on but still not sure how to fix it in a good way. The problem is that for queued outbound frames (in DefaultRemoteFlowController) we will fail the promise but also call `HttpLifecycleHandler.onError(...)` so there is no easy way to understand if it was an "outbound" operation that failed and so the promise will be failed or if it was an inbound operation.  I wonder if we should call `onError(...)` at all for outbound operations.

Because of this we may call `fireExecptionCaught` with such an exception. That said I would love to fix this. @ejona86 any ideas ?

Reply:
also @Scottmitch ^^

Reply:
This may be a good compromise for now https://github.com/netty/netty/pull/7618

Reply:
Just coming around to seeing the replies. Probably would have reduced some confusion if I read things in order :-D.

Yeah, I think that #7618 will improve things here.

Note: as far as I'm concerned, the h2childchan API has no place for Http2Exception. It's unneeded and just makes things confusing. We can still use it "internally," but I feel it is counter-productive to expose to users.

Reply:
@ejona86 hmm... dont you think they would want to receive it sometimes for better visibility what was wrong ?

Reply:
I didn't say there couldn't be exceptions. But why would those exceptions have an `Http2Error`? Why would they have _(shudders)_ a `ShutdownHint`? What is a user supposed to do with such a thing?

The biggest potential problem I see is keeping some distinction between stream- and channel-level exceptions. I think that may not be a big deal for outbound. It could be for inbound, but I also question exactly how many stream-level exceptions we should have for inbound.

Even if we end up keeping Http2Exception for h2childchan, I would hope it becomes less core to the API.

Reply:
> In the case of us receiving a RST_STREAM, I think that should be propagated on the child channel as a userEvent of type Http2ResetFrame.

I'm not very keen on overloading userEvents to carry actual protocol data. Again userEvent is more for control messages between handlers in the pipeline and not intended for protocol  data. Can we treat this the same way would would treat a "half closed" channel, but it isn't optional for HTTP/2?

> Why would they have (shudders) a ShutdownHint

The user is not generally expected to consume this, and doesn't need to be exposed to the user.

> the h2childchan API has no place for Http2Exception

Agreed. For the child channel API we should try to be a consistent as possible with "regular channels" for exceptions and lifecycle.

Reply:
> I'm not very keen on overloading userEvents to carry actual protocol data.

We'll totally need to use userEvents for some protocol data. Anything that is important even when you are pushing-back via flow control (imagine disabled AUTO_READ) seems like it must be a userEvent. This is virtually the same as saying that anything that is useful to reorder in front of data needs to be a userEvent. GoAway is one example. ResetStream is another, since some users will want to stop their processing immediately on receiving of the ResetStream and others will want to drain the inbound data. GoAway and ResetStream are both control messages.

Note that if you ignore the userEvents, you're still okay. You'll still see the channel close after you've read all the data, etc.

Reply:
@Scottmitch @ejona86 I think we should handle inbound `RST_STREAM` as `IOException("connection reset)"` to mimic other protocols and so also support half-closure. Let me do a pr to show the idea.

Reply:
>  I think we should handle inbound RST_STREAM as IOException("connection reset)"

Agreed. Again I think we should be as consistent as possible with other channels which seems like `exceptionCaught` in this case. However in this case the user may also want to get the error code from the RST_STREAM frame. Can/should we put this info in the exception?

> anything that is useful to reorder in front of data needs to be a userEvent.

My main concern is we stay consistent with other channels.
- If there is a corresponding userEvent currently used then we should use it (e.g. stream half closed remote maybe ChannelInputShutdownEvent.INSTANCE](https://github.com/netty/netty/blob/4.1/transport/src/main/java/io/netty/channel/nio/AbstractNioByteChannel.java#L86)).
- If there is an alternative approach used by other channels we should have a good reason to not follow this convention (e.g. RST_STREAM -> exceptionCaught).
- If we exhaust the above options and it makes sense to add a userEvent then lets do it.

I'm not yet convinced that RST_STREAM and GO_AWAY (see https://github.com/netty/netty/pull/7635) necessarily have to be user events.

Reply:
Yeah I think we could just create a custom IOException that hold it... i would like to to do this as a follow up tho

Reply:
followup sgtm!

I would like to get on the same page as @ejona86 too :)

Reply:
Same here... I think these should not be user events

Reply:
> I think we should handle inbound RST_STREAM as IOException("connection reset)" to mimic other protocols and so also support half-closure. Let me do a pr to show the idea.

I was thinking we do something like that as well (if the error code is something other than CANCEL, probably want to include it, though), and that's what most people would use. But receiving RST_STREAM(CANCEL) while pushing-back on flow control seems important for use cases where flow control is likely to be applied, as the notice can arrive much sooner and you can know to abort work without asking for more work.